src/Microsoft.AspNet.Mvc.Core/ModelBinding/CollectionModelBinder.cs

```diff
@@ -150,6 +173,10 @@ public class CollectionModelBinder<TElement> : IModelBinder
                 {
                     didBind = true;
                     boundValue = result.Model;
+                    if (result.ValidationNode != null)
+                    {
+                        validationNode.ChildNodes.Add(result.ValidationNode);
+                    }
```

Comment:
>Loops are also incorrect in that they equate result != null with success. Should check result.IsModelSet for each element. 

Yes that seems to be an existing bug, should be fixed. 

> Gap if the lower-level binder doesn't create a ModelValidationNode but does return a ModelBindingResult. Earlier code unconditionally created a ModelValidationNode for each element that was successfully bound (or perhaps for all elements that seem to exist).

There would be a gap if a lower-level model binder does not create a MVN but creates a result (with model set) where it wanted to validate. However with this design PR there are no such model binders, and with this PR the proposal was to have that contract - If you need validation, create the validation node - 

The alternatives to the apporoch are : 
1. Create a MVN at CompositeModelBinder for any model binding result which is missing it(as we discussed):
                The problem with this approach is it breaks where someone directly tries to point to a particular leaf level model binder (for example using TryUpdateModelBinder to directly point to TypeConverterModelBinder).

2. Create a MVN in MBR by default : 
               ModelBindingResult does not capture ModelMetadata today which means we would have to let every model binder pass the `bindingContext.ModelMetadata` when creating the MBR. 

Lets discuss and reach a conclusion.